processor/batchprocessor/batch_processor.go

```diff
@@ -108,10 +111,19 @@ func (bp *batchProcessor) Start(context.Context, component.Host) error {
 // Shutdown is invoked during service shutdown.
 func (bp *batchProcessor) Shutdown(context.Context) error {
 	bp.cancel()
-	<-bp.done
+
+	// Stop accepting new data.
+	atomic.StoreInt64(&bp.shutdownFlag, 1)
+
+	// Wait until current batch is drained.
+	<-bp.doneCh
```

Comment:
A future iteration could have a timeout so that it doesn't wait indefinitely for this to return.

Reply:
Yes, can be done in a future PR. I want to make sure we first fix the obvious bug.